module/surface_sam_hps.c

```diff
@@ -966,29 +1033,101 @@ static void shps_gpios_remove_irq(struct platform_device *pdev)
 	free_irq(drvdata->irq_dgpu_presence, pdev);
 }
 
+static void shps_sgcp_notify(acpi_handle device, u32 value, void *context) {
+	struct platform_device *pdev = context;
+	switch (value) {
+		case ACPI_SGCP_NOTIFY_POWER_ON:
+			shps_dgpu_powered_on(pdev);
+	}
+}
+
+static int shps_start_sgcp_notification(struct platform_device *pdev, acpi_handle *sgpc_handle) {
+	acpi_handle handle;
+	int status;
+
+	status = acpi_get_handle(NULL, "\\_SB.SGPC", &handle);
+	if (status) {
+		dev_err(&pdev->dev, "error in get_handle %d", status);
+		return status;
+	}
+
+	status = acpi_install_notify_handler(handle, ACPI_DEVICE_NOTIFY, shps_sgcp_notify, pdev);
+	if (status) {
+		dev_err(&pdev->dev, "error in install notify %d", status);
+		*sgpc_handle = NULL;
+		return status;
+	}
+
+	*sgpc_handle = handle; 
+	return 0;
+}
+
+static void shps_remove_sgcp_notification(struct platform_device *pdev) {
+	int status;
+	struct shps_driver_data *drvdata = platform_get_drvdata(pdev);
+
+	if (drvdata->sgpc_handle) {
+		status = acpi_remove_notify_handler(drvdata->sgpc_handle, ACPI_DEVICE_NOTIFY, shps_sgcp_notify);
+		if (status) {
+			dev_err(&pdev->dev, "failed to remove notify handler: %d", status);
+		}
+	}
+}
+
+static struct shps_hardware_traits shps_detect_hardware_traits(struct platform_device *pdev) {
+	const struct shps_hardware_probe *p;
+
+	for (p = shps_hardware_probe_match; p->hardware_id; ++p) {
+		if (acpi_dev_present(p->hardware_id, NULL, -1)) {
+			dev_info(&pdev->dev, 
+				"shps_detect_hardware_traits found device %s, generation %d", 
+				p->hardware_id,
+				p->generation);
+			return *p->hardware_traits;
+		}	
+	}
+	
+	dev_info(
+		&pdev->dev, 
+		"shps_detect_hardware_traits: no specific device found, using defaults");
+	return shps_gen1_hwtraits;
```

Comment:
What I kinda meant here was
```suggestion
	for (p = shps_hardware_probe_match; p->hardware_id; ++p)
		if (acpi_dev_present(p->hardware_id, NULL, -1))
			break;

	dev_info(&pdev->dev,
		 "shps_detect_hardware_traits: using device %s, generation %d\n", 
		 p->hardware_id ? p->hardware_id : "SAN (default)",
		 p->generation);
	
	return *p->hardware_traits;
```
together with replacing the last (NULL) entry in the trait mapping with
```c
	/* Surface Book 2 (default, must be last entry) */
	{ NULL, 1, &shps_gen1_hwtraits },
```
That gets rid of a bit of control flow and also places the defaults in the mapping (so it's a bit more obvious what the defaults are).

Reply:
ahhh ok got it, that makes sense